### PR TITLE
feat: Add GroupHashTombstone

### PR DESCRIPTION
**Pull request by bretthoerner**
Tracks deletion date per GroupHash so that Snuba queries can immediately
exclude deleted groups based on a watershed date.

---

This is just to start tracking the tombstones. Next step will be pushing this down to Snuba but I think it should be broken up into pieces.

### CODE REVIEW

File: src/sentry/models/grouphashtombstone.py
Comment (bretthoerner) [inline src/sentry/models/grouphashtombstone.py]: Definitely open to this living somewhere else under a different name. I didn't want to take over an ORM method like `GroupHash.objects.delete` because there are reasons to delete `GroupHash` without adding tombstones (merge/unmerge).

Also, I'm not sure if I can spare any queries here. It feels a little heavy but nothing is obvious to me.

File: src/sentry/south_migrations/0418_auto__del_grouptagvalue__del_unique_grouptagvalue_group_id_key_value__.py
Comment (bretthoerner) [inline src/sentry/south_migrations/0418_auto__del_grouptagvalue__del_unique_grouptagvalue_group_id_key_value__.py]: So I verified I was based off latest master, and I dropped and recreated my `sentry` tables... but the auto generated migration was full of garbage (like dropping a lot of indexes). I'm going to assume the previously serialized South state is out of sync? I edited down to the new tables I want to add. Is there something I/we should do to fix the state, though?

File: src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py
Comment (bretthoerner) [inline src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py]: @adhiraj Hey I think this is related to your work. I auto-generated this migration and this was included. Was something wrong/missed in a previous migration?

Comment (mattrobenolt) [reply src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py]: @bretthoerner this is intentional, but we’re trying to do a series of like, 5 migrations and it’s taking longer than I anticipated. Do you mind waiting until all of his other bits are merged? Then things will be good. When deleting a column, we have to spread it across a few deploys. And the first step involves removing it from code, without actually deleting the column from the database to avoid errors.

Comment (mattrobenolt) [reply src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py]: I think this is the last bit we need, which is what will conflict here: https://github.com/getsentry/sentry/pull/8615

Comment (bretthoerner) [reply src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py]: Sure, I can update the migration again later. Do you think this is happening today or... longer?

Comment (adhiraj) [reply src/sentry/south_migrations/0421_auto__add_grouphashtombstone__add_unique_grouphashtombstone_project_ha.py]: I'll do it today morning

**Review by tkaemming — APPROVED** [review]

**Pull request merged**
